packages/server/src/unstable-core-do-not-import/http/contentType.ts

```diff
@@ -1,24 +1,91 @@
-import type { TRPCError } from '../error/TRPCError';
-
-export type BodyResult =
-  | {
-      ok: true;
-      data: unknown;
-      /**
-       * If the HTTP handler has already parsed the body
-       */
-      preprocessed: boolean;
+import { TRPCError } from '../error/TRPCError';
+
+type ContentTypeHandler = {
+  isMatch: (opts: Request) => boolean;
+  getInputs: (req: Request, searchParams: URLSearchParams) => Promise<unknown>;
+  batching: boolean;
+  transform: boolean;
+};
+
+const jsonContentTypeHandler: ContentTypeHandler = {
+  async getInputs(req, searchParams) {
+    if (req.method === 'GET') {
+      const input = searchParams.get('input');
+      if (input === null) {
+        return undefined;
+      }
+      return JSON.parse(input);
     }
-  | { ok: false; error: TRPCError };
-
-export type BaseContentTypeHandler<TOptions> = {
-  name: string;
-  isMatch(headers: Headers): boolean;
-  getInputs: (
-    opts: TOptions,
-    info: {
-      isBatchCall: boolean;
-      batch: number;
-    },
-  ) => Promise<unknown>;
+    return await req.json();
+  },
+  isMatch(req) {
+    return !!req.headers.get('content-type')?.startsWith('application/json');
+  },
+  batching: true,
+  transform: true,
+};
+
+const formDataContentTypeHandler: ContentTypeHandler = {
+  async getInputs(req) {
+    if (req.method !== 'POST') {
+      throw new TRPCError({
+        code: 'METHOD_NOT_SUPPORTED',
+        message:
+          'Only POST requests are supported for multipart/form-data requests',
+      });
+    }
+    const fd = await req.formData();
+
+    return fd;
+  },
+  isMatch(req) {
+    return !!req.headers.get('content-type')?.startsWith('multipart/form-data');
+  },
+  batching: false,
+  transform: false,
 };
+
+const octetStreamContentTypeHandler: ContentTypeHandler = {
+  async getInputs(req) {
+    return req.body;
+  },
+  isMatch(req) {
+    return !!req.headers
+      .get('content-type')
+      ?.startsWith('application/octet-stream');
+  },
+  batching: false,
+  transform: false,
+};
+
+const contentTypeHandlers = {
+  list: [
+    formDataContentTypeHandler,
+    jsonContentTypeHandler,
+    octetStreamContentTypeHandler,
+  ],
+  /**
+   * Fallback handler if there is no match
+   */
+  fallback: jsonContentTypeHandler,
+};
+
+export function getContentTypeHandlerOrThrow(req: Request): ContentTypeHandler {
+  const handler = contentTypeHandlers.list.find((handler) =>
+    handler.isMatch(req),
+  );
+  if (handler) {
+    return handler;
+  }
+
+  if (!handler && req.method === 'GET') {
```

Comment:
I removed this behaviour originally because I don't believe it protects against the CSRF attack that we wanted to defend against? Best to throw 405 in any situation that we don't support it?

Reply:
I don't really understand the mechanisms in play though, just read the original issue/PR and got the impression that form verb support is what matters, and forms _do_ have a method="get" option

Reply:
I want to be able to open tRPC GET queries in my browser